File: catalogue/database/circuit_breaker.py
"""
Circuit Breaker Pattern Implementation for Database Operations

This module provides a circuit breaker implementation to prevent cascade failures
when database operations fail repeatedly. It includes fallback mechanisms and
monitoring capabilities.
"""

import asyncio
import logging
from datetime import datetime, timedelta
from enum import Enum
from typing import Any, Callable, Dict, Optional, Union
from dataclasses import dataclass
import threading

logger = logging.getLogger(__name__)


class CircuitBreakerState(Enum):
    """Circuit breaker states."""
    CLOSED = "closed"      # Normal operation
    OPEN = "open"          # Failing, rejecting requests
    HALF_OPEN = "half_open"  # Testing if service recovered


@dataclass
class CircuitBreakerConfig:
    """Configuration for circuit breaker behavior."""
    failure_threshold: int = 5  # Number of failures before opening
    recovery_timeout: int = 60  # Seconds before trying half-open
    success_threshold: int = 3  # Successes needed to close from half-open
    timeout: int = 30  # Operation timeout in seconds


class CircuitBreakerException(Exception):
    """Raised when circuit breaker is open."""
    pass


class DatabaseCircuitBreaker:
    """
    Circuit breaker implementation for database operations.
    
    Prevents cascade failures by monitoring operation success/failure rates
    and temporarily blocking requests when failure threshold is exceeded.
    """
    
    def __init__(self, name: str, config: CircuitBreakerConfig = None):
        self.name = name
        self.config = config or CircuitBreakerConfig()
        self.state = CircuitBreakerState.CLOSED
        self.failure_count = 0
        self.success_count = 0
        self.last_failure_time: Optional[datetime] = None
        self.next_attempt_time: Optional[datetime] = None
        self._lock = threading.Lock()
        
        # Metrics for monitoring
        self.total_requests = 0
        self.total_failures = 0
        self.total_successes = 0
        self.state_changes = []
        
        logger.info(f"Circuit breaker '{name}' initialized with config: {config}")

    def _record_success(self) -> None:
        """Record a successful operation."""
        with self._lock:
            self.total_requests += 1
            self.total_successes += 1
            self.failure_count = 0
            
            if self.state == CircuitBreakerState.HALF_OPEN:
                self.success_count += 1
                if self.success_count >= self.config.success_threshold:
                    self._transition_to_closed()
            
            logger.debug(f"Circuit breaker '{self.name}': Success recorded")

    def _record_failure(self) -> None:
        """Record a failed operation."""
        with self._lock:
            self.total_requests += 1
            self.total_failures += 1
            self.failure_count += 1
            self.success_count = 0
            self.last_failure_time = datetime.utcnow()
            
            if (self.state == CircuitBreakerState.CLOSED and 
                self.failure_count >= self.config.failure_threshold):
                self._transition_to_open()
            elif self.state == CircuitBreakerState.HALF_OPEN:
                self._transition_to_open()
            
            logger.warning(
                f"Circuit breaker '{self.name}': Failure recorded "
                f"(count: {self.failure_count})"
            )

    def _transition_to_open(self) -> None:
        """Transition circuit breaker to OPEN state."""
        self.state = CircuitBreakerState.OPEN
        self.next_attempt_time = (
            datetime.utcnow() + timedelta(seconds=self.config.recovery_timeout)
        )
        self._record_state_change("OPEN")
        logger.warning(
            f"Circuit breaker '{self.name}' opened. "
            f"Next attempt at {self.next_attempt_time}"
        )

    def _transition_to_half_open(self) -> None:
        """Transition circuit breaker to HALF_OPEN state."""
        self.state = CircuitBreakerState.HALF_OPEN
        self.success_count = 0
        self._record_state_change("HALF_OPEN")
        logger.info(f"Circuit breaker '{self.name}' half-opened for testing")

    def _transition_to_closed(self) -> None:
        """Transition circuit breaker to CLOSED state."""
        self.state = CircuitBreakerState.CLOSED
        self.failure_count = 0
        self.success_count = 0
        self.next_attempt_time = None
        self._record_state_change("CLOSED")
        logger.info(f"Circuit breaker '{self.name}' closed - normal operation resumed")

    def _record_state_change(self, new_state: str) -> None:
        """Record state change for monitoring."""
        self.state_changes.append({
            "timestamp": datetime.utcnow().isoformat(),
            "state": new_state,
            "failure_count": self.failure_count,
            "total_failures": self.total_failures
        })
        
        # Keep only last 100 state changes
        if len(self.state_changes) > 100:
            self.state_changes = self.state_changes[-100:]

    def _can_attempt(self) -> bool:
        """Check if an operation attempt is allowed."""
        if self.state == CircuitBreakerState.CLOSED:
            return True
        elif self.state == CircuitBreakerState.OPEN:
            if (self.next_attempt_time and 
                datetime.utcnow() >= self.next_attempt_time):
                self._transition_to_half_open()
                return True
            return False
        elif self.state == CircuitBreakerState.HALF_OPEN:
            return True
        return False

    async def call(self, operation: Callable, *args, **kwargs) -> Any:
        """
        Execute an operation through the circuit breaker.
        
        Args:
            operation: The async function to execute
            *args: Arguments for the operation
            **kwargs: Keyword arguments for the operation
            
        Returns:
            Result of the operation
            
        Raises:
            CircuitBreakerException: When circuit breaker is open
            Exception: Any exception from the operation
        """
        if not self._can_attempt():
            raise CircuitBreakerException(
                f"Circuit breaker '{self.name}' is OPEN. "
                f"Next attempt allowed at {self.next_attempt_time}"
            )

        try:
            # Execute operation with timeout
            result = await asyncio.wait_for(
                operation(*args, **kwargs),
                timeout=self.config.timeout
            )
            self._record_success()
            return result
            
        except asyncio.TimeoutError as e:
            self._record_failure()
            logger.error(f"Operation timeout in circuit breaker '{self.name}': {e}")
            raise
            
        except Exception as e:
            self._record_failure()
            logger.error(f"Operation failed in circuit breaker '{self.name}': {e}")
            raise

    def get_metrics(self) -> Dict[str, Any]:
        """
        Get circuit breaker metrics for monitoring.
        
        Returns:
            Dictionary containing current metrics
        """
        with self._lock:
            failure_rate = (
                (self.total_failures / self.total_requests * 100)
                if self.total_requests > 0 else 0
            )
            
            return {
                "name": self.name,
                "state": self.state.value,
                "failure_count": self.failure_count,
                "success_count": self.success_count,
                "total_requests": self.total_requests,
                "total_failures": self.total_failures,
                "total_successes": self.total_successes,
                "failure_rate_percent": round(failure_rate, 2),
                "last_failure_time": (
                    self.last_failure_time.isoformat() 
                    if self.last_failure_time else None
                ),
                "next_attempt_time": (
                    self.next_attempt_time.isoformat() 
                    if self.next_attempt_time else None
                ),
                "config": {
                    "failure_threshold": self.config.failure_threshold,
                    "recovery_timeout": self.config.recovery_timeout,
                    "success_threshold": self.config.success_threshold,
                    "timeout": self.config.timeout
                },
                "recent_state_changes": self.state_changes[-10:]  # Last 10 changes
            }

    def reset(self) -> None:
        """Reset circuit breaker to initial state."""
        with self._lock:
            self.state = CircuitBreakerState.CLOSED
            self.failure_count = 0
            self.success_count = 0
            self.last_failure_time = None
            self.next_attempt_time = None
            logger.info(f"Circuit breaker '{self.name}' reset to initial state")


class DatabaseFallbackHandler:
    """
    Handles fallback mechanisms when database operations fail.
    """
    
    def __init__(self):
        self.cache = {}  # Simple in-memory cache for fallback data
        
    async def get_fallback_products(self) -> list:
        """
        Return fallback product data when database is unavailable.
        
        Returns:
            List of basic product information
        """
        logger.info("Returning fallback product data")
        return [
            {
                "id": "fallback-1",
                "name": "Basic Origami Paper",
                "description": "Standard origami paper for beginners",
                "image_url": "/static/images/origami/001-origami.png",
                "price": 9.99,
                "category": "paper",
                "active": True,
                "featured": False
            },
            {
                "id": "fallback-2", 
                "name": "Origami Instruction Book",
                "description": "Learn basic origami techniques",
                "image_url": "/static/images/origami/002-book.png",
                "price": 15.99,
                "category": "books",
                "active": True,
                "featured": True
            }
        ]
    
    async def get_fallback_product(self, product_id: str) -> Optional[dict]:
        """
        Return fallback data for a specific product.
        
        Args:
            product_id: ID of the requested product
            
        Returns:
            Product data or None if not available
        """
        fallback_products = await self.get_fallback_products()
        for product in fallback_products:
            if product["id"] == product_id:
                return product
        return None
    
    def cache_successful_response(self, key: str, data: Any, ttl: int = 300) -> None:
        """
        Cache successful responses for fallback use.
        
        Args:
            key: Cache key
            data: Data to cache
            ttl: Time to live in seconds
        """
        expiry = datetime.utcnow() + timedelta(seconds=ttl)
        self.cache[key] = {
            "data": data,
            "expiry": expiry
        }
        
        # Clean expired entries
        self._clean_expired_cache()
    
    def get_cached_response(self, key: str) -> Optional[Any]:
        """
        Get cached response if available and not expired.
        
        Args:
            key: Cache key
            
        Returns:
            Cached data or None
        """
        if key in self.cache:
            entry = self.cache[key]
            if datetime.utcnow() < entry["expiry"]:
                return entry["data"]
            else:
                del self.cache[key]
        return None
    
    def _clean_expired_cache(self) -> None:
        """Remove expired cache entries."""
        now = datetime.utcnow()
        expired_keys = [
            key for key, entry in self.cache.items()
            if now >= entry["expiry"]
        ]
        for key in expired_keys:
            del self.cache[key]


# Global circuit breaker instances
_circuit_breakers: Dict[str, DatabaseCircuitBreaker] = {}
_fallback_handler = DatabaseFallbackHandler()


def get_circuit_breaker(name: str, config: CircuitBreakerConfig = None) -> DatabaseCircuitBreaker:
    """
    Get or create a circuit breaker instance.
    
    Args:
        name: Circuit breaker name
        config: Configuration (uses default if not provided)
        
    Returns:
        Circuit breaker instance
    """
    if name not in _circuit_breakers:
        _circuit_breakers[name] = DatabaseCircuitBreaker(name, config)
    return _circuit_breakers[name]


def get_fallback_handler() -> DatabaseFallbackHandler:
    """Get the global fallback handler instance."""
    return _fallback_handler


def get_all_circuit_breaker_metrics() -> Dict[str, Any]:
    """
    Get metrics for all circuit breakers.
    
    Returns:
        Dictionary with metrics for each circuit breaker
    """
    return {
        name: cb.get_metrics() 
        for name, cb in _circuit_breakers.items()
    }

File: recommendation/database/circuit_breaker.go
package database

import (
	"context"
	"fmt"
	"log"
	"sync"
	"time"
)

// CircuitBreakerState represents the current state of the circuit breaker
type CircuitBreakerState int

const (
	// StateClosed - Normal operation, requests are allowed
	StateClosed CircuitBreakerState = iota
	// StateOpen - Failing, requests are rejected
	StateOpen
	// StateHalfOpen - Testing if service recovered
	StateHalfOpen
)

// String returns the string representation of the circuit breaker state
func (s CircuitBreakerState) String() string {
	switch s {
	case StateClosed:
		return "CLOSED"
	case StateOpen:
		return "OPEN"
	case StateHalfOpen:
		return "HALF_OPEN"
	default:
		return "UNKNOWN"
	}
}

// CircuitBreakerConfig holds the configuration for circuit breaker behavior
type CircuitBreakerConfig struct {
	FailureThreshold   int           // Number of failures before opening
	RecoveryTimeout    time.Duration // Time before trying half-open
	SuccessThreshold   int           // Successes needed to close from half-open
	Timeout           time.Duration // Operation timeout
	MaxRequests       int           // Max requests allowed in half-open state
}

// DefaultCircuitBreakerConfig returns a default configuration
func DefaultCircuitBreakerConfig() *CircuitBreakerConfig {
	return &CircuitBreakerConfig{
		FailureThreshold: 5,
		RecoveryTimeout:  60 * time.Second,
		SuccessThreshold: 3,
		Timeout:         30 * time.Second,
		MaxRequests:     3,
	}
}

// CircuitBreakerError represents an error when circuit breaker is open
type CircuitBreakerError struct {
	Message string
	State   CircuitBreakerState
}

func (e *CircuitBreakerError) Error() string {
	return e.Message
}

// StateChange represents a circuit breaker state change event
type StateChange struct {
	Timestamp    time.Time `json:"timestamp"`
	FromState    string    `json:"from_state"`
	ToState      string    `json:"to_state"`
	FailureCount int       `json:"failure_count"`
	TotalFailures int      `json:"total_failures"`
}

// Metrics holds circuit breaker metrics for monitoring
type Metrics struct {
	Name              string         `json:"name"`
	State             string         `json:"state"`
	FailureCount      int            `json:"failure_count"`
	SuccessCount      int            `json:"success_count"`
	TotalRequests     int64          `json:"total_requests"`
	TotalFailures     int64          `json:"total_failures"`
	TotalSuccesses    int64          `json:"total_successes"`
	FailureRate       float64        `json:"failure_rate_percent"`
	LastFailureTime   *time.Time     `json:"last_failure_time,omitempty"`
	NextAttemptTime   *time.Time     `json:"next_attempt_time,omitempty"`
	Config            *CircuitBreakerConfig `json:"config"`
	RecentStateChanges []StateChange `json:"recent_state_changes"`
}

// CircuitBreaker implements the circuit breaker pattern for database operations
type CircuitBreaker struct {
	name            string
	config          *CircuitBreakerConfig
	state           CircuitBreakerState
	failureCount    int
	successCount    int
	totalRequests   int64
	totalFailures   int64
	totalSuccesses  int64
	lastFailureTime *time.Time
	nextAttemptTime *time.Time
	stateChanges    []StateChange
	mutex           sync.RWMutex
}

// NewCircuitBreaker creates a new circuit breaker instance
func NewCircuitBreaker(name string, config *CircuitBreakerConfig) *CircuitBreaker {
	if config == nil {
		config = DefaultCircuitBreakerConfig()
	}

	cb := &CircuitBreaker{
		name:         name,
		config:       config,
		state:        StateClosed,
		stateChanges: make([]StateChange, 0),
	}

	log.Printf("Circuit breaker '%s' initialized with config: %+v", name, config)
	return cb
}

// Execute runs an operation through the circuit breaker
func (cb *CircuitBreaker) Execute(ctx context.Context, operation func(context.Context) (interface{}, error)) (interface{}, error) {
	if !cb.canAttempt() {
		return nil, &CircuitBreakerError{
			Message: fmt.Sprintf("Circuit breaker '%s' is %s. Next attempt allowed at %v",
				cb.name, cb.state.String(), cb.nextAttemptTime),
			State: cb.state,
		}
	}

	// Create context with timeout
	timeoutCtx, cancel := context.WithTimeout(ctx, cb.config.Timeout)
	defer cancel()

	// Execute operation
	result, err := operation(timeoutCtx)

	if err != nil {
		cb.recordFailure()
		return nil, err
	}

	cb.recordSuccess()
	return result, nil
}

// canAttempt checks if an operation attempt is allowed
func (cb *CircuitBreaker) canAttempt() bool {
	cb.mutex.Lock()
	defer cb.mutex.Unlock()

	switch cb.state {
	case StateClosed:
		return true
	case StateOpen:
		if cb.nextAttemptTime != nil && time.Now().After(*cb.nextAttemptTime) {
			cb.transitionToHalfOpen()
			return true
		}
		return false
	case StateHalfOpen:
		return true
	}
	return false
}

// recordSuccess records a successful operation
func (cb *CircuitBreaker) recordSuccess() {
	cb.mutex.Lock()
	defer cb.mutex.Unlock()

	cb.totalRequests++
	cb.totalSuccesses++
	cb.failureCount = 0

	if cb.state == StateHalfOpen {
		cb.successCount++
		if cb.successCount >= cb.config.SuccessThreshold {
			cb.transitionToClosed()
		}
	}

	log.Printf("Circuit breaker '%s': Success recorded", cb.name)
}

// recordFailure records a failed operation
func (cb *CircuitBreaker) recordFailure() {
	cb.mutex.Lock()
	defer cb.mutex.Unlock()

	cb.totalRequests++
	cb.totalFailures++
	cb.failureCount++
	cb.successCount = 0
	now := time.Now()
	cb.lastFailureTime = &now

	if (cb.state == StateClosed && cb.failureCount >= cb.config.FailureThreshold) ||
		cb.state == StateHalfOpen {
		cb.transitionToOpen()
	}

	log.Printf("Circuit breaker '%s': Failure recorded (count: %d)", cb.name, cb.failureCount)
}

// transitionToOpen transitions circuit breaker to OPEN state
func (cb *CircuitBreaker) transitionToOpen() {
	oldState := cb.state
	cb.state = StateOpen
	nextAttempt := time.Now().Add(cb.config.RecoveryTimeout)
	cb.nextAttemptTime = &nextAttempt
	cb.recordStateChange(oldState, StateOpen)
	log.Printf("Circuit breaker '%s' opened. Next attempt at %v", cb.name, nextAttempt)
}

// transitionToHalfOpen transitions circuit breaker to HALF_OPEN state
func (cb *CircuitBreaker) transitionToHalfOpen() {
	oldState := cb.state
	cb.state = StateHalfOpen
	cb.successCount = 0
	cb.recordStateChange(oldState, StateHalfOpen)
	log.Printf("Circuit breaker '%s' half-opened for testing", cb.name)
}

// transitionToClosed transitions circuit breaker to CLOSED state
func (cb *CircuitBreaker) transitionToClosed() {
	oldState := cb.state
	cb.state = StateClosed
	cb.failureCount = 0
	cb.successCount = 0
	cb.nextAttemptTime = nil
	cb.recordStateChange(oldState, StateClosed)
	log.Printf("Circuit breaker '%s' closed - normal operation resumed", cb.name)
}

// recordStateChange records a state change for monitoring
func (cb *CircuitBreaker) recordStateChange(fromState, toState CircuitBreakerState) {
	change := StateChange{
		Timestamp:     time.Now(),
		FromState:     fromState.String(),
		ToState:       toState.String(),
		FailureCount:  cb.failureCount,
		TotalFailures: int(cb.totalFailures),
	}

	cb.stateChanges = append(cb.stateChanges, change)

	// Keep only last 100 state changes
	if len(cb.stateChanges) > 100 {
		cb.stateChanges = cb.stateChanges[len(cb.stateChanges)-100:]
	}
}

// GetMetrics returns current circuit breaker metrics
func (cb *CircuitBreaker) GetMetrics() *Metrics {
	cb.mutex.RLock()
	defer cb.mutex.RUnlock()

	var failureRate float64
	if cb.totalRequests > 0 {
		failureRate = float64(cb.totalFailures) / float64(cb.totalRequests) * 100
	}

	// Get recent state changes (last 10)
	recentChanges := make([]StateChange, 0)
	if len(cb.stateChanges) > 0 {
		start := len(cb.stateChanges) - 10
		if start < 0 {
			start = 0
		}
		recentChanges = cb.stateChanges[start:]
	}

	return &Metrics{
		Name:               cb.name,
		State:              cb.state.String(),
		FailureCount:       cb.failureCount,
		SuccessCount:       cb.successCount,
		TotalRequests:      cb.totalRequests,
		TotalFailures:      cb.totalFailures,
		TotalSuccesses:     cb.totalSuccesses,
		FailureRate:        failureRate,
		LastFailureTime:    cb.lastFailureTime,
		NextAttemptTime:    cb.nextAttemptTime,
		Config:             cb.config,
		RecentStateChanges: recentChanges,
	}
}

// Reset resets the circuit breaker to initial state
func (cb *CircuitBreaker) Reset() {
	cb.mutex.Lock()
	defer cb.mutex.Unlock()

	cb.state = StateClosed
	cb.failureCount = 0
	cb.successCount = 0
	cb.lastFailureTime = nil
	cb.nextAttemptTime = nil
	log.Printf("Circuit breaker '%s' reset to initial state", cb.name)
}

// GetState returns the current state of the circuit breaker
func (cb *CircuitBreaker) GetState() CircuitBreakerState {
	cb.mutex.RLock()
	defer cb.mutex.RUnlock()
	return cb.state
}

// IsOpen returns true if the circuit breaker is open
func (cb *CircuitBreaker) IsOpen() bool {
	return cb.GetState() == StateOpen
}

// FallbackHandler provides fallback mechanisms when operations fail
type FallbackHandler struct {
	cache map[string]CacheEntry
	mutex sync.RWMutex
}

// CacheEntry represents a cached entry with expiration
type CacheEntry struct {
	Data      interface{}
	ExpiresAt time.Time
}

// NewFallbackHandler creates a new fallback handler
func NewFallbackHandler() *FallbackHandler {
	return &FallbackHandler{
		cache: make(map[string]CacheEntry),
	}
}

// GetFallbackRecommendations returns fallback recommendation data
func (fh *FallbackHandler) GetFallbackRecommendations() []OrigamiRecommendation {
	log.Println("Returning fallback recommendation data")
	
	return []OrigamiRecommendation{
		{
			ID:          "fallback-1",
			Name:        "Paper Crane",
			Description: "Classic origami crane - symbol of peace and hope",
			ImageURL:    "/static/images/origami/007-dove.png",
			Score:       0.95,
			Reason:      "Popular beginner choice",
		},
		{
			ID:          "fallback-2",
			Name:        "Paper Butterfly",
			Description: "Beautiful butterfly design for beginners",
			ImageURL:    "/static/images/origami/008-butterfly.png",
			Score:       0.90,
			Reason:      "Colorful and engaging",
		},
		{
			ID:          "fallback-3",
			Name:        "Paper Elephant",
			Description: "Cute elephant design with simple folds",
			ImageURL:    "/static/images/origami/004-elephant.png",
			Score:       0.85,
			Reason:      "Fun animal design",
		},
	}
}

// GetFallbackOrigamiOfTheDay returns fallback origami of the day
func (fh *FallbackHandler) GetFallbackOrigamiOfTheDay() OrigamiRecommendation {
	recommendations := fh.GetFallbackRecommendations()
	if len(recommendations) > 0 {
		return recommendations[0] // Return first as default
	}
	
	return OrigamiRecommendation{
		ID:          "default",
		Name:        "Simple Paper Fold",
		Description: "Basic origami for beginners",
		ImageURL:    "/static/images/origami/001-origami.png",
		Score:       0.80,
		Reason:      "Default recommendation",
	}
}

// CacheSuccessfulResponse caches a successful response for fallback use
func (fh *FallbackHandler) CacheSuccessfulResponse(key string, data interface{}, ttl time.Duration) {
	fh.mutex.Lock()
	defer fh.mutex.Unlock()

	fh.cache[key] = CacheEntry{
		Data:      data,
		ExpiresAt: time.Now().Add(ttl),
	}

	// Clean expired entries
	fh.cleanExpiredCache()
}

// GetCachedResponse retrieves a cached response if available and not expired
func (fh *FallbackHandler) GetCachedResponse(key string) (interface{}, bool) {
	fh.mutex.RLock()
	defer fh.mutex.RUnlock()

	entry, exists := fh.cache[key]
	if !exists {
		return nil, false
	}

	if time.Now().After(entry.ExpiresAt) {
		delete(fh.cache, key)
		return nil, false
	}

	return entry.Data, true
}

// cleanExpiredCache removes expired cache entries
func (fh *FallbackHandler) cleanExpiredCache() {
	now := time.Now()
	for key, entry := range fh.cache {
		if now.After(entry.ExpiresAt) {
			delete(fh.cache, key)
		}
	}
}

// Global instances
var (
	circuitBreakers = make(map[string]*CircuitBreaker)
	fallbackHandler = NewFallbackHandler()
	cbMutex         sync.RWMutex
)

// GetCircuitBreaker gets or creates a circuit breaker instance
func GetCircuitBreaker(name string, config *CircuitBreakerConfig) *CircuitBreaker {
	cbMutex.Lock()
	defer cbMutex.Unlock()

	if cb, exists := circuitBreakers[name]; exists {
		return cb
	}

	cb := NewCircuitBreaker(name, config)
	circuitBreakers[name] = cb
	return cb
}

// GetFallbackHandler returns the global fallback handler instance
func GetFallbackHandler() *FallbackHandler {
	return fallbackHandler
}

// GetAllCircuitBreakerMetrics returns metrics for all circuit breakers
func GetAllCircuitBreakerMetrics() map[string]*Metrics {
	cbMutex.RLock()
	defer cbMutex.RUnlock()

	metrics := make(map[string]*Metrics)
	for name, cb := range circuitBreakers {
		metrics[name] = cb.GetMetrics()
	}
	return metrics
}